Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+    consuming func writeAndFinish(_ buffer: ByteBuffer) async throws
+    /// Report an error in the response stream
+    consuming func reportError(_ error: any Error) async throws
+}
+```
+
+### LambdaContext
+
+`LambdaContext` will be largely unchanged, but the `eventLoop` property as well as the `allocator` property (of type `ByteBuffer`) will be removed.
+
+A new function `backgroundTask()` will also be added. This will allow tasks to be run in the background while and after the response is/has been sent. Please note that `LambdaContext` will not be Sendable anymore.
+
+```swift
+/// A context object passed as part of an invocation in LambdaHandler handle functions.
+public struct LambdaContext {
+    /// The request ID, which identifies the request that triggered the function invocation.
+    public var requestID: String { get }
+
+    /// The AWS X-Ray tracing header.
+    public var traceID: String { get }
+
+    /// The ARN of the Lambda function, version, or alias that's specified in the invocation.
+    public var invokedFunctionARN: String { get }
+
+    /// The timestamp that the function times out.
+    public var deadline: DispatchWallTime { get }
+
+    /// For invocations from the AWS Mobile SDK, data about the Amazon Cognito identity provider.
+    public var cognitoIdentity: String? { get }
+
+    /// For invocations from the AWS Mobile SDK, data about the client application and device.
+    public var clientContext: String? { get }
+
+    /// `Logger` to log with.
+    ///
+    /// - note: The `LogLevel` can be configured using the `LOG_LEVEL` environment variable.
+    public var logger: Logger { get }
+
+    /// Schedule background work in lambda. This work must be completed before
+    /// the runtime can ask for further events. Note that AWS will continue to
+    /// charge you after the response has been returned, but work is still
+    /// processing in the background.
+    public func addBackgroundTask(_ body: sending @escaping () async -> ())
+}
+```
+
+### StreamingLambdaHandler
+
+The new `StreamingLambdaHandler` protocol is the base protocol to implement a Lambda function. Most users will not use this protocol and instead use the `LambdaHandler` protocol defined in the `Codable` Support section.
+
+```swift
+/// The base LambdaHandler protocol
+public protocol StreamingLambdaHandler: ~Copyable {
+    /// The business logic of the Lambda function
+    /// - Parameters:
+    ///   - event: The invocation's input data
+    ///   - responseWriter: A ``LambdaResponseWriter`` to write the invocations response to.
+    ///                     If no response or error is written to the `responseWriter` it will
+    ///                     report an error to the invoker.
+    ///   - context: The LambdaContext containing the invocation's metadata
+    mutating func handle(_ event: ByteBuffer, responseWriter: consuming some LambdaResponseWriter, context: LambdaContext) async
```

Comment:
Shouldn't we make it possible that this method throws? Otherwise it forces every user to wrap any throwing calls into a do/catch. FWIW in [grpc](https://github.com/grpc/grpc-swift/blob/main/Sources/GRPC/Docs.docc/Proposals/0001-stub-api.md) we did exactly the opposite and handle that for the user since it makes it way more ergonomic to write those handlers

Reply:
Yeah I think I can get behind that.

Reply:
As mentioned above: https://github.com/swift-server/swift-aws-lambda-runtime/pull/339#discussion_r1711150410

You convinced me that this should `throw`. 

- If an error is thrown before the first write happened, we will use the error endpoint. 
- If an error is thrown after the write has completed, we will log the error.
- If an error is thrown while sending the response we use trailing headers to report the error.

@tachyonics What we hand to the user as a Non-Copyable type can just be a wrapper around what we use internally :)